### PR TITLE
Wrapping remaining AppContext methods + Files API

### PR DESCRIPTION
**Pull request by jguz-pubnub**

### CODE REVIEW

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/files/ListFiles.ios.kt [hunk 1]
```diff
@@ -1,9 +1,54 @@
...
+        pubnub.listFilesWithChannel(
+            channel = channel,
+            limit = limit?.let { NSNumber(it) },
+            next = createPubNubHashedPage(from = next),
+            onSuccess = callback.onSuccessHandler2 { data, next ->
+                PNListFilesResult(
+                    count = data?.size ?: 0, // TODO: count property is not retrieved from ListFilesSuccessResponse in Swift SDK
+                    next = next?.let { PNPage.PNNext(pageHash = it) },
+                    status = 200,
+                    data = filteredList(from = data) { rawValue: PubNubFileObjC ->
```
Comment (marcin-cebo) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/files/ListFiles.ios.kt:41]: For me it is a bit misleading that `data` is used for two purpose here. First as lambda param and second as property of PNListFilesResult. What about 
`onSuccess = callback.onSuccessHandler2 { input, nextInput->`
?

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/GetChannelMembers.ios.kt [hunk 2]
```diff
@@ -1,9 +1,64 @@
...
+    override fun async(callback: Consumer<Result<PNMemberArrayResult>>) {
+        pubnub.getChannelMembersWithChannel(
+            channel = channel,
+            limit = limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeUUIDFields = includeUUIDDetails == PNUUIDDetailsLevel.UUID,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/GetChannelMembers.ios.kt:50]: shouldn't it be `PNUUIDDetailsLevel.UUID || PNUUIDDetailsLevel.UUID_WITH_CUSTOM` ?

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt [hunk 3]
```diff
@@ -1,9 +1,106 @@
...
+        pubnub.setChannelMembersWithChannel(
+            channel = channel,
+            uuids = uuids.map { PubNubUUIDMetadataObjC(id = it.uuid, custom = it.custom?.value, status = it.status) },
+            limit = limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeUUIDFields = includeUUIDDetails == PNUUIDDetailsLevel.UUID,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt:53]: ditto

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt [hunk 4]
```diff
@@ -1,9 +1,106 @@
...
+        pubnub.removeChannelMembersWithChannel(
+            channel = channel,
+            uuids = uuids,
+            limit = limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeUUIDFields = includeUUIDDetails == PNUUIDDetailsLevel.UUID,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt:92]: and here

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/GetMemberships.ios.kt [hunk 5]
```diff
@@ -1,9 +1,62 @@
...
+    override fun async(callback: Consumer<Result<PNChannelMembershipArrayResult>>) {
+        pubnub.getMembershipsWithUuid(
+            uuid = uuid,
+            limit =  limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeChannelFields = includeChannelDetails == PNChannelDetailsLevel.CHANNEL,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/GetMemberships.ios.kt:48]: same situation

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/ManageMemberships.ios.kt [hunk 6]
```diff
@@ -1,9 +1,105 @@
...
+        pubnub.setMembershipsWithChannels(
+            channels = channels,
+            uuid = uuid,
+            limit = limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeChannelFields = includeChannelDetails == PNChannelDetailsLevel.CHANNEL,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/ManageMemberships.ios.kt:52]: and here

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/ManageMemberships.ios.kt [hunk 7]
```diff
@@ -1,9 +1,105 @@
...
+        pubnub.removeMembershipsWithChannels(
+            channels = channels,
+            uuid = uuid,
+            limit = limit?.let { NSNumber(it) },
+            page = createPubNubHashedPage(from = page),
+            filter = filter,
+            sort = sort.map { it.key.fieldName },
+            includeCount = includeCount,
+            includeCustom = includeCustom,
+            includeChannelFields = includeChannelDetails == PNChannelDetailsLevel.CHANNEL,
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/membership/ManageMemberships.ios.kt:91]: and here

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt [hunk 8]
```diff
@@ -1,10 +1,113 @@
...
 import platform.Foundation.NSString
 import platform.Foundation.NSUTF8StringEncoding
 import platform.Foundation.dataUsingEncoding
 
 fun String.toNSData(): NSData? {
     return (this as NSString).dataUsingEncoding(NSUTF8StringEncoding)
+}
+
+@OptIn(ExperimentalForeignApi::class)
+fun createPubNubHashedPage(from: PNPage?): PubNubHashedPageObjC {
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt:26]: if these functions are not meant to be public you could make them `internal`

**Review by marcin-cebo — APPROVED** [review]

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt [hunk 9]
```diff
@@ -1,10 +1,113 @@
...
+    return PNMember(
+        uuid = from.uuid()?.let { createPNUUIDMetadata(from = it) },
+        custom = from.custom() as Map<String, Any>, // TODO: Check
+        updated = from.updated() ?: "",
+        eTag = from.eTag() ?: "",
+        status = from.status()
+    )
+}
+
+inline fun <reified T, U> filteredList(from: List<*>?, mapper: (T) -> U): Collection<U> {
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt:105]: I'd probably make it an extension function so you can chain it easily:

```inline fun <reified T, U> List<*>?.filteredList(mapper: (T) -> U): Collection<U> {
    return this?.filterIsInstance<T>()?.map { mapper(it) } ?: emptyList()
}```

**Review by wkal-pubnub — CHANGES_REQUESTED** [review]

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt (hunk 9 above)
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt:105]: also `map { mapper(it) }` is the same as `map(mapper)`

Comment (jguz-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt:105]: Alright, I'm not sure if the name of the function I proposed is the best one. What I want to do is:

1. Ensuring that a collection contains `T` elements
2. If yes, transform each element into `U` by calling `((T) -> `U)`

What about naming it like `transform`?

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/files/ListFiles.ios.kt (hunk 1 above)
Comment (jguz-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/files/ListFiles.ios.kt:41]: Sure, I'll change parameter names from the lambda expression:

```
onSuccess = callback.onSuccessHandler2 { files, nextPageHash
```

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt (hunk 9 above)
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt]: how about `filterAndMap` ?

Comment (jguz-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/kmp/Converters.kt]: Sounds good, done!

File: pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt (hunk 4 above)
Comment (wkal-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt:92]: I think you made a mistake in this one, check this and line below (`||` is in wrong line)

Comment (jguz-pubnub) [reply pubnub-kotlin/pubnub-kotlin-api/src/iosMain/kotlin/com/pubnub/api/endpoints/objects/member/ManageChannelMembers.ios.kt]: Indeed, thx for catching this. Fixed

**Review by wkal-pubnub — APPROVED** [review]

**Review by marcin-cebo — APPROVED** [review]

**Pull request merged**